File: src/main/java/dominio/indicadores/ExpresionOperacion.java
package dominio.indicadores;

import dominio.empresas.Empresa;

import java.time.Year;

public class ExpresionOperacion implements Expresion{
	private Expresion operandoIzq;
	private Expresion operandoDer;
	private OperacionAritmetica operadorAritmetico;
	
	public ExpresionOperacion(Expresion operandoIzq, Expresion operandoDer, String operadorBinario){
		this.operandoIzq=operandoIzq;
		this.operandoDer=operandoDer;
		if(operadorBinario.equals("+")) this.operadorAritmetico = OperacionAritmetica.Mas;
		else if (operadorBinario.equals("-")) this.operadorAritmetico = OperacionAritmetica.Menos;
		else if (operadorBinario.equals("*")) this.operadorAritmetico = OperacionAritmetica.Por;
		else if (operadorBinario.equals("/")) this.operadorAritmetico = OperacionAritmetica.Dividido;
	}
	
	public int evaluarEn(Empresa empresa, Year anio) {
		return operadorAritmetico.applyAsInt(operandoIzq.evaluarEn(empresa, anio), operandoDer.evaluarEn(empresa, anio));
	}
	
}

File: src/main/java/dondeInvierto/dominio/indicadores/ExpresionOperacion.kt
package dondeInvierto.dominio.indicadores

import dondeInvierto.dominio.empresas.Empresa
import java.lang.IllegalArgumentException

import java.time.Year

class ExpresionOperacion(private val operandoIzq: Expresion,
                         private val operandoDer: Expresion,
                         operadorBinario: String) : Expresion {

    private val operadorAritmetico: OperacionAritmetica = when (operadorBinario) {
        "+" -> OperacionAritmetica.Mas
        "-" -> OperacionAritmetica.Menos
        "*" -> OperacionAritmetica.Por
        "/" -> OperacionAritmetica.Dividido
        else -> throw IllegalArgumentException("Las únicas operaciones aceptadas son: [+,-,*,/]")
    }

    override fun evaluarEn(empresa: Empresa, anio: Year): Int {
        return operadorAritmetico.applyAsInt(operandoIzq.evaluarEn(empresa, anio), operandoDer.evaluarEn(empresa, anio))
    }

}
